File: tasks/python/task-02-billing-month-end/solution/billing.py
"""Billing dates for monthly subscriptions."""

import calendar
from datetime import date


def add_months(start, months):
    """Return the date `months` after `start`.

    The day stays the same when it exists. Otherwise it is the last day
    of the target month.
    """
    month_index = start.month - 1 + months
    year = start.year + month_index // 12
    month = month_index % 12 + 1
    last_day = calendar.monthrange(year, month)[1]
    day = min(start.day, last_day)
    return date(year, month, day)


def next_billing_date(start, today):
    """Return the first billing date that is on or after `today`."""
    billing = start
    months = 0
    while billing < today:
        months += 1
        billing = add_months(start, months)
    return billing


def billing_dates(start, count):
    """Return the first `count` billing dates, starting with `start`."""
    return [add_months(start, i) for i in range(count)]


def days_until_billing(start, today):
    """Return how many days remain until the next billing date."""
    return (next_billing_date(start, today) - today).days

File: tasks/typescript/task-02-billing-month-end/solution/billing.ts
/**
 * Billing dates for monthly subscriptions.
 *
 * Every date is a Date at midnight UTC, for example new Date("2024-01-31").
 */

const DAY_MS = 24 * 60 * 60 * 1000;

/**
 * Return the date `months` after `start`.
 *
 * The day stays the same when it exists. Otherwise it is the last day
 * of the target month.
 */
export function addMonths(start: Date, months: number): Date {
  const monthIndex = start.getUTCMonth() + months;
  const year = start.getUTCFullYear() + Math.floor(monthIndex / 12);
  const month = monthIndex % 12;
  const lastDay = new Date(Date.UTC(year, month + 1, 0)).getUTCDate();
  const day = Math.min(start.getUTCDate(), lastDay);
  return new Date(Date.UTC(year, month, day));
}

/** Return the first billing date that is on or after `today`. */
export function nextBillingDate(start: Date, today: Date): Date {
  let billing = start;
  let months = 0;
  while (billing < today) {
    months += 1;
    billing = addMonths(start, months);
  }
  return billing;
}

/** Return the first `count` billing dates, starting with `start`. */
export function billingDates(start: Date, count: number): Date[] {
  return Array.from({ length: count }, (_, i) => addMonths(start, i));
}

/** Return how many days remain until the next billing date. */
export function daysUntilBilling(start: Date, today: Date): number {
  const next = nextBillingDate(start, today);
  return Math.round((next.getTime() - today.getTime()) / DAY_MS);
}
